src/core/init/manifest_update_scheduler.ts

```diff
@@ -20,103 +20,187 @@ import {
- * @param {Object} initialManifest
- * @param {Observable} scheduleRefresh$
- * @param {Function} fetchManifest
- * @param {number} minimumManifestUpdateInterval
+ * @param {Object} manifestUpdateSchedulerArguments
  * @returns {Observable}
  */
-export default function manifestUpdateScheduler(
-  initialManifest : { manifest : Manifest;
-                      sendingTime? : number; },
-  scheduleRefresh$ : Observable<number>,
-  fetchManifest : IManifestFetcher,
-  minimumManifestUpdateInterval : number
-) : Observable<never> {
+export default function manifestUpdateScheduler({
+  fetchManifest,
+  initialManifest,
+  manifestUpdateUrl,
+  maximumManifestUpdateInterval,
+  minimumManifestUpdateInterval,
+  scheduleRefresh$,
+} : IManifestUpdateSchedulerArguments) : Observable<never> {
+  // The Manifest always keeps the same Manifest
+  const { manifest } = initialManifest;
+
   function handleManifestRefresh$(
     manifestInfos: { manifest: Manifest;
-                     sendingTime?: number; }): Observable<never> {
-    const { manifest, sendingTime } = manifestInfos;
-    // schedule a Manifest refresh to avoid sending too much request.
-    const manualRefresh$ = scheduleRefresh$.pipe(
-      mergeMap((delay) => {
-        const timeSinceLastRefresh = sendingTime == null ?
-                                       0 :
-                                       performance.now() - sendingTime;
-        const minInterval = Math.max(minimumManifestUpdateInterval
-                                       - timeSinceLastRefresh,
-                                     0);
-        return observableTimer(Math.max(delay - timeSinceLastRefresh,
-                                        minInterval));
-      }));
-
-    const autoRefresh$ = (() => {
-      if (manifest.lifetime == null || manifest.lifetime <= 0) {
-        return EMPTY;
+                     sendingTime?: number;
+                     receivedTime? : number;
+                     parsingTime : number;
+                     updatingTime? : number; }): Observable<never> {
+    const { sendingTime } = manifestInfos;
+
+    const internalRefresh$ = scheduleRefresh$.pipe(mergeMap(({ completeRefresh }) =>
+      startInternalRefreshInterval(completeRefresh,
+                                   minimumManifestUpdateInterval,
+                                   sendingTime).pipe(mapTo({ completeRefresh }))));
+
+    const timeSinceRequest = sendingTime == null ? 0 :
+                                                   performance.now() - sendingTime;
+    const minInterval = Math.max(minimumManifestUpdateInterval - timeSinceRequest, 0);
+
+    let autoRefresh$;
+    if (manifest.lifetime === undefined || manifest.lifetime < 0) {
+      autoRefresh$ = EMPTY;
+    } else {
+      const { parsingTime, updatingTime } = manifestInfos;
+      let autoRefreshInterval = manifest.lifetime * 1000 - timeSinceRequest;
+      if (parsingTime + (updatingTime ?? 0) >= manifest.lifetime / 4) {
```

Comment:
/!\ Manifest lifetime is in seconds, while parsing time is in milliseconds.
We should test again that optim.

Reply:
Yep, done